### PR TITLE
[RFC] hello_world: add demo code for using custom ree service

### PR DESCRIPTION
**Pull request by divneil**
Signed-off-by: Divneil Rai Wadhawan <divneil.r.wadhawan@intel.com>

### CODE REVIEW

File: hello_world/host/main.c [hunk 1]
```diff
@@ -28,21 +28,92 @@
...
 	TEEC_Context ctx;
 	TEEC_Session sess;
 	TEEC_Operation op;
+	pthread_t ree_serv_thread;
 	TEEC_UUID uuid = TA_HELLO_WORLD_UUID;
 	uint32_t err_origin;
+	void *service;
+	REEC_UUID ree_uuid = TA_HELLO_WORLD_REE_UUID;
+
+#if 0
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply hello_world/host/main.c]: What's dynlib? Is msgq Posix Message Queues?

[the remark this answers is not in the document]
Comment (divneil) [reply hello_world/host/main.c]: dynlib -> dynamic library (.so). 
msgq -> Yes Posix Message Queue.

File: hello_world/host/main.c [hunk 2]
```diff
@@ -101,6 +172,11 @@ int main(void)
 	TEEC_CloseSession(&sess);
 
 	TEEC_FinalizeContext(&ctx);
+#if 0
```
[the remark this answers is not in the document]
Comment (divneil) [reply hello_world/host/main.c]: I will create separate flows for message queues and dynamic lib, so, both will be enabled by default.

**Comment by divneil** [comment]
> Another suggestion here would be to create a new folder `dynlib` (or some other clever name) and 
I thought of creating a new folder `ta2ca` which will extend `hello_world`. To copy the dynamic library to file system, can you please share which folder I need to use. Considering the process I need to follow to share the code, it will take some time.

**Comment by jbech-linaro** [comment]
> To copy the dynamic library to file system, can you please share which folder I need to use

I'm not too picky about that, try to come up with a name of a folder that has something to do with the example host/ta that you're trying to implement. You can see the rationale behind the ones we already have here: https://optee.readthedocs.io/en/latest/building/gits/optee_examples/optee_examples.html#example-applications

But the concept is the same in all for them.
``` bash
./aes/ta
./aes/host
./hello_world
./hello_world/ta
./hello_world/host
....
```
If you need to add other code/libraries to fulfill the implementation of your example, then put it under "your" example folder. I've said a lot, did I manage to answer your question? :) 

Btw, when you're code has been merged, I'd kindly ask you to also submit a documentation update for things like this.

**Comment by divneil** [comment]
> I'm not too picky about that, try to come up with a name of a folder that has something to do with 

Honestly, I am unable to come up with a small catchy name. I will try with `ta2ca` (Trusted Application to Client Application), but feel free to suggest any other name.

Hmm, I will rephrase the question. Earlier, I was pretty abstract. 
Once, we build the distro, we get `rootfs.cpio.gz`. I am unable to make `.so` go into this `cpio` from `host` folder. 

So, if I need push a patch in let's say `build` folder (or any folder other than optee_os/optee_client/optee_examples) to fix this, I will need an approval from my organization, so, wanted that process to start in parallel. So, can you please help with the git I need to target.

> Btw, when you're code has been merged, I'd kindly ask you to also submit a documentation update for things like this.

Definitely.

File: hello_ree/host/main.c [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+ *   to fulfill the CA request. The REE service could be storage or generically,
+ *   anything not supported by tee-supplicant.
+ *
+ * o This application uses two available solutions to provide service to TEE.
+ *   It is based on using a UUID in the reverse direction (UTA -> CA)
+ *   - Posix Message Queues: Message Queue is transparent to the caller.
+ *
+ *   - Dynamic Library     : The library needs to be developed with a specific
+ *                           function signature and compiled as .so and placed
+ *                           in a PATH (/lib64) where tee-supplicant can find
```
Comment (etienne-lms) [inline hello_ree/host/main.c]: suggest to replace `/lib64` with `i.e. /lib64` or even `/lib`.

File: hello_ree/host/main.c [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+ * with permission to access /data folder, otherwise, invocation will fail and
+ * TA will receive a failure.
+ */
+
+#include <err.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <pthread.h>
```
Comment (etienne-lms) [inline hello_ree/host/main.c]: minor: sort in alpha ordering

File: hello_ree/ta/Makefile [hunk 5]
```diff
@@ -0,0 +1,13 @@
+CFG_TEE_TA_LOG_LEVEL ?= 4
```
Comment (etienne-lms) [inline hello_ree/ta/Makefile]: minor: maybe 2 or 3 as default trace level.

File: hello_ree/ta/hello_ree_ta.c [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+/**
+ * invoke_ree_service() - call the REE service
+ */
+static TEE_Result invoke_ree_service(TEE_UUID *uuid)
+{
+	char msg_to_ree[] = "Hello! from TEE";
+	ree_session_handle ree_sess = NULL;
+	TEE_Result result = TEE_SUCCESS;
+	TEE_Param ree_params[4] = {0};
+	uint32_t param_types;
```
Comment (etienne-lms) [inline hello_ree/ta/hello_ree_ta.c]: prefer always initialize local variables where defined, here add `= 0`.

File: hello_ree/ta/include/hello_ree_ta.h [hunk 7]
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (C) 2019 Intel Corporation All Rights Reserved
+ */
+
+#ifndef __TA_HELLO_REE_H__
+#define __TA_HELLO_REE_H__
```
Comment (etienne-lms) [inline hello_ree/ta/include/hello_ree_ta.h]: `__HELLO_REE_TA_H__` as per filename.

**Pull request closed without merging**